### PR TITLE
Add softplus activation function.

### PR DESCRIPTION
**Pull request by nowozin**
Adding the [_softplus_ activation function](https://en.wikipedia.org/wiki/Rectifier_%28neural_networks%29), named in [(Dugas et al., NIPS 2000)](http://papers.nips.cc/paper/1920-incorporating-second-order-functional-knowledge-for-better-option-pricing.pdf), defined as f(x)=log(1+exp(x)).

One note: I may have misunderstood how the pairs of functions and their gradients in `mshadow_op.h` are supposed to be used.  It seems there are two use cases which differ:

1. Use case one: e.g. `tanh`, `sigmoid`, and now `softplus` have gradient functions which are not actually their gradient but instead the functions that when evaluated on the output of the basic operator result provides the gradient.  For example, if `a=tanh(x)` then `g=tanh_grad(a)` is the gradient of `tanh` at x.
2. Use case two: e.g. `log` has a gradient `log_grad` which follows the canonical definition of the gradient, i.e. 1/x.

It may be worthwhile to replace the `_grad` suffix where needed with something that indicates the first use, for example `_gradfromresult` or similar.

### CODE REVIEW

File: src/operator/activation.cu [hunk 1]
```diff
@@ -14,6 +14,10 @@ namespace mxnet {
 namespace op {
 template<>
 Operator *CreateOp<gpu>(ActivationParam param) {
+  // Not supported by CUDNN yet
+  if (param.act_type == activation::kSoftplus)
+      return new ActivationOp<gpu, mshadow_op::softplus, mshadow_op::softplus_grad>();
+
```
Comment (piiswrong) [inline src/operator/activation.cu]: this is unnecessary. we want it to go into cudnn-activation and report an error.

File: src/operator/activation-inl.h [hunk 2]
```diff
@@ -24,7 +24,7 @@ namespace op {
 namespace activation {
 enum ActivationOpInputs {kData};
 enum ActivationOpOutputs {kOut};
-enum ActivationOpType {kReLU, kSigmoid, kTanh};
+enum ActivationOpType {kReLU, kSigmoid, kTanh, kSoftplus};
```
Comment (piiswrong) [inline src/operator/activation-inl.h]: These days it's more commonly known as SoftReLU I think.
Consider change the naming

**Comment by piiswrong** [comment]
Hi nowozin, Thanks for the contribution. I made a few comments. Otherwise looks good to me.

As to the confusion about grad functions. I think it's a fair point. @tqchen @antinucleon  what do you think.

File: src/operator/activation-inl.h (hunk 2 above)
Comment (nowozin) [inline src/operator/activation-inl.h]: Thanks for pointing this out.  Agreed, I have changed the name to `SoftReLU` and `softrelu`.

File: src/operator/activation.cu (hunk 1 above)
Comment (nowozin) [inline src/operator/activation.cu]: Are you sure?  If I remove this check a FATAL error will be thrown by the `CuDNNActivationOp` constructor.  In contrast, the current patch works with and without CUDNN support compiled.

**Comment by piiswrong** [comment]
Fair point. I think the current solution is fine.

**Comment by tqchen** [comment]
Thanks for the contribution, I agree that adding gradient from result is more clear, please feel free to open a PR to do so. 

I have added you to https://github.com/dmlc/mxnet/blob/master/CONTRIBUTORS.md Please feel free to update more information about yourself and contribute your wisdom to improving mxnet

**Pull request merged**